Variorium Connector: kernel granularity and JSON

twilk10 opened the pull request:
Updates to the current connector with updates such as json updates, energy estimation, device information. It has also been altered to get power information at the kernel level.

Commit:
first commit number 3

Commit:
variorum updates

Commit:
variorum updates

masterleinad commented:
Please explain the behavior of the connector tool including options. What is the minimum variorum version required? We should check for that when trying to find `Variorum` via `CMake`.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -18,25 +19,34 @@
+#include <stdio.h>
+#include <string>
+#include <unordered_set>
+#include <vector>
```
Which of these header files are actually required? Please make sure to only use what you need.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -18,25 +19,34 @@
+uint32_t gdevID = -1;
+std::vector<float> gpu_powers;
+std::vector<float> gpu_powers2;
+std::string output = "";
```
What are all these global variables used for? Please add comments in the code explaining their usage.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -74,10 +84,39 @@ std::string printpath = "./";
+    0; // 0 is for both print power & json, 1 is for print power, 2 is for json
+bool usingMPI = false;
+bool verbosePrint = true;
+bool mpiOutPut = false;
```
Are all of these still used/necessary?

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -74,10 +84,39 @@ std::string printpath = "./";
+        std::cout << "File created or opened successfully." << std::endl;
+    }
+    file.close(); // Close the file
+i}*/
```
Please remove unused code.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -74,10 +84,39 @@ std::string printpath = "./";
+    file.close(); // Close the file
+i}*/
+void printFile() {
+  const std::string filename = "variorumoutput.txt";
```
Maybe make this configurable via an environment variable.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -74,10 +84,39 @@ std::string printpath = "./";
+    }
+    file.close(); // Close the file
+i}*/
+void printFile() {
```
```suggestion
void createFile() {
```

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -74,10 +84,39 @@ std::string printpath = "./";
+    std::cout << "File does not exist. Creating a new file." << std::endl;
+    std::ofstream file(filename, std::ios::app);
+  }
+  infile.close();
```
```suggestion
```
Doesn't this happen we the object is destroyed anyway?

masterleinad replied:
ping

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -88,20 +127,22 @@ bool mpiOutPut    = false;
+  const char *hostnameChar =
       json_string_value(json_object_get(power_obj, "hostname"));
-  std::string hostname(hostnameChar);
+  // std::string hostname(hostnameChar);
```
```suggestion
```

masterleinad replied:
ping

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -119,7 +160,7 @@ std::string variorum_print_power_call() {
         json_real_value(json_object_get(power_obj, "power_gpu_socket_1"));
 
-    outputString += "HostName " + hostname + "\n";
+    // outputString += "HostName " + hostname + "\n";
```
```suggestion
```

masterleinad replied:
ping

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -130,7 +171,7 @@ std::string variorum_print_power_call() {
     outputString += "\n GPU Socket 2: " + std::to_string(power_gpu1) + "\n";
   } else {
-    outputString += hostname + ": " + std::to_string(power_node) + "\n";
+    // outputString += hostname + ": " + std::to_string(power_node) + "\n";
```
```suggestion
```

masterleinad replied:
ping

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -88,20 +127,22 @@ bool mpiOutPut    = false;
-  std::string hostname(hostnameChar);
+  // std::string hostname(hostnameChar);
   // print informatin to screen
   if (verbosePrint) {
```
Is this used at all or at least useful?

masterleinad replied:
Just remove

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -169,7 +210,7 @@ void variorum_call_mpi() {
     std::ofstream pfile;
     if (type_of_profiling ==
-        0) {  // if both print power and json options selected
+        0) { // if both print power and json options selected
```
Is this still useful?

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -226,27 +267,138 @@ void variorum_call_mpi() {
-
 void variorum_call() {
-  std::string output;
+  // std::cout <<output  << std::endl;
```
```suggestion
```

masterleinad replied:
ping

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -226,27 +267,138 @@ void variorum_call_mpi() {
-    std::cout << output << std::endl;
+    output = variorum_print_power_call();
+    char *s = variorum_json_call();
+    // std::cout << s << "\n";
```
```suggestion
```

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -226,27 +267,138 @@ void variorum_call_mpi() {
+    output = variorum_print_power_call();
+    char *s = variorum_json_call();
+    // std::cout << s << "\n";
+    char *timestamp_pos = strstr(s, "\"timestamp\":");
+    if (timestamp_pos != nullptr) {
+      timestamp_pos += strlen("\"timestamp\":");
+      char *timestamp_end_pos = strchr(timestamp_pos, ',');
+      if (timestamp_end_pos == nullptr) {
+        timestamp_end_pos = strchr(timestamp_pos, '}');
+      }
+      if (timestamp_end_pos != nullptr) {
+        std::string timestamp_str(timestamp_pos,
+                                  timestamp_end_pos - timestamp_pos);
+        long long timestamp = std::stoll(timestamp_str);
+        // std::cout << "Timestamp: " << timestamp << std::endl;
+        if (power == -1 || time2 != 0) {
+          time1 = timestamp;
+          // std::cout << ""
+        } else {
+          time2 = timestamp;
+        }
+      }
```
Can't you use jansson for parsing the json output?
Please add comments on what this code is doing.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -226,27 +267,138 @@ void variorum_call_mpi() {
+                                    num_sockets_end_pos - num_sockets_pos);
+
+        num_sockets = std::stoll(num_sockets_str);
+        //          std::cout << "Number of Sockets: " << num_sockets <<
+        //          std::endl;
```
```suggestion
```

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -226,27 +267,138 @@ void variorum_call_mpi() {
+        }
+      }
+    }
+    char *num_sockets_pos = strstr(s, "\"num_gpus_per_socket\":");
+    if (num_sockets_pos != nullptr) {
+      num_sockets_pos += strlen("\"num_gpus_per_socket\":");
+
+      char *num_sockets_end_pos = strchr(num_sockets_pos, ',');
+      if (num_sockets_end_pos == nullptr) {
+        num_sockets_end_pos = strchr(num_sockets_pos, '}');
+      }
+      if (num_sockets_end_pos != nullptr) {
+
+        std::string num_sockets_str(num_sockets_pos,
+                                    num_sockets_end_pos - num_sockets_pos);
+
+        num_sockets = std::stoll(num_sockets_str);
```
Please try using jansson here as well (and below).

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -226,27 +267,138 @@ void variorum_call_mpi() {
+      }
+    }
+    if (power != 0) {
+      std::cout << "  Gpu start power: " << gpu_powers2[gdevID]
+                << " Gpu end power: " << gpu_powers[gdevID]
+                << "   Energy Estimation:   "
+                << ((gpu_powers[gdevID] + gpu_powers2[0]) / 2) *
+                       ((temp - time1) * .001)
+                << std::endl;
```
Let's just not print anything to stdout but only to the file.

vlkale replied:
Sounds fine to me.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -357,25 +509,102 @@ void kokkosp_init_library(const int loadSeq, const uint64_t interfaceVer,
-    variorum_call_mpi();
-  } else {
-    variorum_call();
+  /*	std::cout << "kokkos library call\n" << std::endl;
+    if (usingMPI) {
+      variorum_call_mpi();
+    } else {
+     variorum_call();
+    }
+    time_t total_time;
+    time_t end_time;
+    time(&end_time);
+    std::cout << "End Time: " << end_time << "\nStart Time: " << start_time
+              << "\n";
+    total_time = end_time - start_time;
+
+    std::cout << "The kokkos library was alive for " << total_time << "
+    seconds."
+              << std::endl;*/
```
```suggestion
```

vlkale replied:
@twilk10  Is this suggestion OK with you? If so, can you commit it? 

It is good with me.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -357,25 +509,102 @@ void kokkosp_init_library(const int loadSeq, const uint64_t interfaceVer,
-void kokkosp_begin_parallel_for(const char* name, const uint32_t devID,
-                                uint64_t* kID) {
-  std::cout << "Device ID: " << devID << "\n";
+std::ostream &operator<<(
+    std::ostream &os,
+    const Kokkos::Tools::Experimental::ExecutionSpaceIdentifier &identifier) {
+
+  os << "  Type: " << identifier.type << "";
+  os << "  Device ID: " << identifier.device_id << "";
+  gdevID = identifier.device_id;
+  os << "  Instance ID: " << identifier.instance_id;
+  //   output += " Device ID: " + std::to_string(identifier.device_id) +
+  //                      " Instance ID: " +
+  //                      std::to_string(identifier.instance_id);
+  writeToFile("variorumoutput.txt", output);
+  return os;
+}
```
Don't overload `operator<<` but just write to the output file directly.

masterleinad reviewed (commented):
Please provide sample output.

masterleinad commented:
Please use `clang-format 8` for the indentation.

vlkale commented:
> Please explain the behavior of the connector tool including options. What is the minimum variorum version required? We should check for that when trying to find `Variorum` via `CMake`.

+1

 @twilk10 Thanks for submitting this PR!

To provide an explanation of  this connector tool’s options and a high-level behavior from the Kokkos user perspective: maybe you can add `kokkosp_print_help()` (like what is in space-time-stack).

twilk10 commented:
> > Please explain the behavior of the connector tool including options. What is the minimum variorum version required? We should check for that when trying to find `Variorum` via `CMake`.
> 
> +1
> 
> @twilk10 Thanks for submitting this PR!
> 
> To provide an explanation of this connector tool’s options and a high-level behavior from the Kokkos user perspective: maybe you can add `kokkosp_print_help()` (like what is in space-time-stack).

Thank you I will work on getting that information up as soon as possible.

vlkale commented:
> > > Please explain the behavior of the connector tool including options. What is the minimum variorum version required? We should check for that when trying to find `Variorum` via `CMake`.
> > 
> > 
> > +1
> > @twilk10 Thanks for submitting this PR!
> > To provide an explanation of this connector tool’s options and a high-level behavior from the Kokkos user perspective: maybe you can add `kokkosp_print_help()` (like what is in space-time-stack).
> 
> Thank you I will work on getting that information up as soon as possible.



> > > Please explain the behavior of the connector tool including options. What is the minimum variorum version required? We should check for that when trying to find `Variorum` via `CMake`.
> > 
> > 
> > +1
> > @twilk10 Thanks for submitting this PR!
> > To provide an explanation of this connector tool’s options and a high-level behavior from the Kokkos user perspective: maybe you can add `kokkosp_print_help()` (like what is in space-time-stack).
> 
> Thank you I will work on getting that information up as soon as possible.

Thanks! Be sure to run clang-format-8 on each code file that you changed in this PR, as @masterleinad mentions in one of his comments. This is a prereq for your PR to be merged. Note that this is the CI check that doesn't pass.

twilk10 commented:
> > > > Please explain the behavior of the connector tool including options. What is the minimum variorum version required? We should check for that when trying to find `Variorum` via `CMake`.
> > > 
> > > 
> > > +1
> > > @twilk10 Thanks for submitting this PR!
> > > To provide an explanation of this connector tool’s options and a high-level behavior from the Kokkos user perspective: maybe you can add `kokkosp_print_help()` (like what is in space-time-stack).
> > 
> > 
> > Thank you I will work on getting that information up as soon as possible.
> 
> > > > Please explain the behavior of the connector tool including options. What is the minimum variorum version required? We should check for that when trying to find `Variorum` via `CMake`.
> > > 
> > > 
> > > +1
> > > @twilk10 Thanks for submitting this PR!
> > > To provide an explanation of this connector tool’s options and a high-level behavior from the Kokkos user perspective: maybe you can add `kokkosp_print_help()` (like what is in space-time-stack).
> > 
> > 
> > Thank you I will work on getting that information up as soon as possible.
> 
> Thanks! Be sure to run clang-format-8 on each code file that you changed in this PR, as @masterleinad mentions in one of his comments. This is a prereq for your PR to be merged. Note that this is the CI check that doesn't pass.

I ran clang-format 8 on it I will have to evaluate what went wrong and try it again.

Commit:
clang update

vlkale commented:
> > > > > Please explain the behavior of the connector tool including options. What is the minimum variorum version required? We should check for that when trying to find `Variorum` via `CMake`.
> > > > 
> > > > 
> > > > +1
> > > > @twilk10 Thanks for submitting this PR!
> > > > To provide an explanation of this connector tool’s options and a high-level behavior from the Kokkos user perspective: maybe you can add `kokkosp_print_help()` (like what is in space-time-stack).
> > > 
> > > 
> > > Thank you I will work on getting that information up as soon as possible.
> > 
> > 
> > > > > Please explain the behavior of the connector tool including options. What is the minimum variorum version required? We should check for that when trying to find `Variorum` via `CMake`.
> > > > 
> > > > 
> > > > +1
> > > > @twilk10 Thanks for submitting this PR!
> > > > To provide an explanation of this connector tool’s options and a high-level behavior from the Kokkos user perspective: maybe you can add `kokkosp_print_help()` (like what is in space-time-stack).
> > > 
> > > 
> > > Thank you I will work on getting that information up as soon as possible.
> > 
> > 
> > Thanks! Be sure to run clang-format-8 on each code file that you changed in this PR, as @masterleinad mentions in one of his comments. This is a prereq for your PR to be merged. Note that this is the CI check that doesn't pass.
> 
> I ran clang-format 8 on it I will have to evaluate what went wrong and try it again.

@twilk10 OK. You want to do something like: 

`clang-format-8 -style=file --assume-filename=../../.clang-format ../../profiling/variorium-connector/kp_variorium_connector.cpp > ../../profiling/variorium-connector/kp_variorium_connector-temp.cpp `

You can diff in between to see whether and/or how the file has indeed changed and then `mv` the temporary file to the original file. 

If that doesn't work, can you send the error output with the flag `-Werror` ?

vlkale commented:
@twilk10 Can you look at @masterleinad 's comments particularly on printing output and parsing json with jansson? I think they are good suggestions and maybe you can commit those before committing other changes with clang-format.

Also, make sure to note to not include any unneeded headers.

twilk10 commented:
> @twilk10 Can you look at @masterleinad 's comments particularly on printing output and parsing json with jansson? I think they are good suggestions and maybe you can commit those before committing other changes with clang-format.
> 
> Also, make sure to note to not include any unneeded headers.

I will take a look and try updating withthe help of these suggestions.

twilk10 commented:
> > > > > > Please explain the behavior of the connector tool including options. What is the minimum variorum version required? We should check for that when trying to find `Variorum` via `CMake`.
> > > > > 
> > > > > 
> > > > > +1
> > > > > @twilk10 Thanks for submitting this PR!
> > > > > To provide an explanation of this connector tool’s options and a high-level behavior from the Kokkos user perspective: maybe you can add `kokkosp_print_help()` (like what is in space-time-stack).
> > > > 
> > > > 
> > > > Thank you I will work on getting that information up as soon as possible.
> > > 
> > > 
> > > > > > Please explain the behavior of the connector tool including options. What is the minimum variorum version required? We should check for that when trying to find `Variorum` via `CMake`.
> > > > > 
> > > > > 
> > > > > +1
> > > > > @twilk10 Thanks for submitting this PR!
> > > > > To provide an explanation of this connector tool’s options and a high-level behavior from the Kokkos user perspective: maybe you can add `kokkosp_print_help()` (like what is in space-time-stack).
> > > > 
> > > > 
> > > > Thank you I will work on getting that information up as soon as possible.
> > > 
> > > 
> > > Thanks! Be sure to run clang-format-8 on each code file that you changed in this PR, as @masterleinad mentions in one of his comments. This is a prereq for your PR to be merged. Note that this is the CI check that doesn't pass.
> > 
> > 
> > I ran clang-format 8 on it I will have to evaluate what went wrong and try it again.
> 
> @twilk10 OK. You want to do something like:
> 
> `clang-format-8 -style=file --assume-filename=../../.clang-format ../../profiling/variorium-connector/kp_variorium_connector.cpp > ../../profiling/variorium-connector/kp_variorium_connector-temp.cpp `
> 
> You can diff in between to see whether and/or how the file has indeed changed and then `mv` the temporary file to the original file.
> 
> If that doesn't work, can you send the error output with the flag `-Werror` ?

I am having trouble getting access to clang-format-8 but I can get other versions.

twilk10 commented:
> > > > > > > Please explain the behavior of the connector tool including options. What is the minimum variorum version required? We should check for that when trying to find `Variorum` via `CMake`.
> > > > > > 
> > > > > > 
> > > > > > +1
> > > > > > @twilk10 Thanks for submitting this PR!
> > > > > > To provide an explanation of this connector tool’s options and a high-level behavior from the Kokkos user perspective: maybe you can add `kokkosp_print_help()` (like what is in space-time-stack).
> > > > > 
> > > > > 
> > > > > Thank you I will work on getting that information up as soon as possible.
> > > > 
> > > > 
> > > > > > > Please explain the behavior of the connector tool including options. What is the minimum variorum version required? We should check for that when trying to find `Variorum` via `CMake`.
> > > > > > 
> > > > > > 
> > > > > > +1
> > > > > > @twilk10 Thanks for submitting this PR!
> > > > > > To provide an explanation of this connector tool’s options and a high-level behavior from the Kokkos user perspective: maybe you can add `kokkosp_print_help()` (like what is in space-time-stack).
> > > > > 
> > > > > 
> > > > > Thank you I will work on getting that information up as soon as possible.
> > > > 
> > > > 
> > > > Thanks! Be sure to run clang-format-8 on each code file that you changed in this PR, as @masterleinad mentions in one of his comments. This is a prereq for your PR to be merged. Note that this is the CI check that doesn't pass.
> > > 
> > > 
> > > I ran clang-format 8 on it I will have to evaluate what went wrong and try it again.
> > 
> > 
> > @twilk10 OK. You want to do something like:
> > `clang-format-8 -style=file --assume-filename=../../.clang-format ../../profiling/variorium-connector/kp_variorium_connector.cpp > ../../profiling/variorium-connector/kp_variorium_connector-temp.cpp `
> > You can diff in between to see whether and/or how the file has indeed changed and then `mv` the temporary file to the original file.
> > If that doesn't work, can you send the error output with the flag `-Werror` ?
> 
> I am having trouble getting access to clang-format-8 but I can get other versions.

I have solved this issue.

vlkale commented:
Great! All checks have passed for clang-format-8. Can you address the comments by @masterleinad ? (You can commit suggestions if you agree). 

I also updated the title to capture the changes you made in your code better. 

@masterleinad Are there any other things to resolve here?

twilk10 commented:
> Great! All checks have passed for clang-format-8. Can you address the comments by @masterleinad ? (You can commit suggestions if you agree).
> 
> I also updated the title to capture the changes you made in your code better.
> 
> @masterleinad Are there any other things to resolve here?

I am currently working on updating the connector to address the comments made by @masterleinad

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -1,4 +1,5 @@
-//@HEADER
+//@HEADER140
```
```suggestion
//@HEADER
```

masterleinad replied:
ping

masterleinad replied:
You still have `HEADER140` here.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -357,25 +508,99 @@ void kokkosp_init_library(const int loadSeq, const uint64_t interfaceVer,
+    case Kokkos::Tools::Experimental::DeviceType::HIP:
+      os << "hip";
+      output += "Type: HIP";
+      break;
+    case Kokkos::Tools::Experimental::DeviceType::OpenMPTarget:
+      os << "openmptarget";
+      output += "Type: openmptarget";
+      break;
+    case Kokkos::Tools::Experimental::DeviceType::HPX:
+      os << "hpx";
+      output += "Type: hpx";
+      break;
+    case Kokkos::Tools::Experimental::DeviceType::Threads:
+      os << "threads";
+      output += "Type: threads";
+      break;
+    case Kokkos::Tools::Experimental::DeviceType::SYCL:
+      os << "sycl";
+      output += "Type: SYCL";
+      break;
+    case Kokkos::Tools::Experimental::DeviceType::OpenACC:
+      os << "openacc";
+      output += "Type: OPENACC";
+      break;
+
+    default:
+      os << "Unknown Device Type";
+      output += "Type: Uknown Device Type";
+      break;
   }
-  time_t total_time;
-  time_t end_time;
-  time(&end_time);
-  std::cout << "End Time: " << end_time << "\nStart Time: " << start_time
-            << "\n";
-  total_time = end_time - start_time;
-
-  std::cout << "The kokkos library was alive for " << total_time << " seconds."
-            << std::endl;
+  return os;
```
Define a function such as https://github.com/kokkos/kokkos-tools/blob/5e895a36f40bfce363c1424a6f5529a9ef6b64de/profiling/space-time-stack/kp_space_time_stack.cpp#L54-L67 instead. We don't want to make `Kokkos::Tools::Experimental::DeviceType` implicitly printable.

Commit:
updated based on concerns

twilk10 committed:
Merge branch 'kokkos:develop' into varconnect

twilk10 commented:
This is an update to the Kokkos-tool variorum connector and can be used to get the power data an
Tested on incremental HIP test on exascale machine using AMD style architecture.
The idea of change is to make this tool usable and change the functionality to retrieve power information at the kernel level as that is a more modern use for the tool.
This is a step in the direction of allowing users to measure their energy consumption while using Kokkos. 
Features to be updated
	-updated json parsing 
	-updated what is to be printed out 
Features that were updated in variorum but others
	-API to get power 
Features implemented
	-Equation to estimate energy usage until energy API is updated
	-Moving the capture of data to the hooks so that the power 	data being retrieved is from inside the kernel
Outputs include device Id device type device kernel and power alongside power estimation
Also Outputs to a file that is made on the system. User can provide an environment variable to be used by uncommenting code but if not variorum will make a file for the user to put the information in. 

Compatible with variorum version 6.0
Sample output;
```name: Z4mainEUllRlE_ Device ID: 0 Instance ID: 1 DeviceType: SERIAL  Energy Estimation 2679.104000```

Rombur commented:
You pushed your build directory. You need to remove it.

Commit:
remove build directory

masterleinad commented on profiling/variorum-connector/kp_variorium_connector-temp.cpp:
Remove this file.

masterleinad commented on profiling/variorum-connector/kp_variorum_connector-temp.cpp:
Remove this file.

masterleinad commented on profiling/variorum-connector/varconnectsample:
Remove this file.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -26,13 +27,23 @@
 #include <regex>
-#include <cxxabi.h>
-#include <dlfcn.h>
+//#include <cxxabi.h>
+//#include <dlfcn.h>
```
```suggestion
```
If you don't need it, remove it.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -26,13 +27,23 @@
 #include <iostream>
 #include <fstream>
-
+int power       = -1;
+int power1      = 0;
+int power2      = 0;
+int filemake    = -1;
+long long time1 = 0;
+long long time2 = 0;
+uint32_t gdevID = -1;
+std::vector<float> gpu_powers;
+std::vector<float> gpu_powers2;
+std::string output   = "";
+std::string filename = "variorumoutput.txt";
```
Please add comments for these variables.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -75,9 +86,50 @@ time_t start_time;
-bool usingMPI     = false;
-bool verbosePrint = false;
-bool mpiOutPut    = false;
+// bool usingMPI     = false;
+bool verbosePrint = true;
+// bool mpiOutPut    = false;
```
Remove what you don't use.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -75,9 +86,50 @@ time_t start_time;
+    //   kokkosp_print_help(argv[0]);
+    exit(1);
+  }
+}
```
Do you need this?
```suggestion
```

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -75,9 +86,50 @@ time_t start_time;
+  variable!\n"; std::abort();
+    }
+  std::string filename = parsed_output_file;
+  */
```
Please remove all commented code.
```suggestion
```

masterleinad replied:
What about inlining this function?

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -75,9 +86,50 @@ time_t start_time;
+  if (file) {
+    file << data;
+  }
+}
```
What about inlining this function?

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -142,107 +195,201 @@ std::string variorum_print_power_call() {
+      gpu_powers.clear();
+      gpu_powers2.clear();
+    }
+    // std::cout<< s << std::endl;
```
```suggestion
```

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -142,107 +195,201 @@ std::string variorum_print_power_call() {
+    json_t* value;
+    json_object_foreach(root, key, value) {
+      if (json_is_object(value)) {
+        json_t* socket_object = json_object_get(value, "socket_0");
```
Why is `socket_0` special?

twilk10 replied:
In the json object socket_0 is the structure in which the gpu powers are within.

masterleinad replied:
What if there are multiple sockets?

twilk10 replied:
I have not seen multiple sockets when testing so I have not considered it do you know any way that there would be multiple sockets for the Json object and how that would change the object?

masterleinad replied:
I am seeing multiple sockets on Intel testbeds: https://github.com/LLNL/variorum/pull/563#issue-2380895406
Just add a comment like
```
// FIXME We assume that all GPUs are on socket 0 for now.
```

Commit:
varconnect deleted

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -142,107 +195,201 @@ std::string variorum_print_power_call() {
+    if (!json_is_object(power_gpu_watts)) {
+      fprintf(stderr, "Expected 'power_gpu_watts' to be an object.\n");
+    }
+    if (gdevID == 0) {
```
Why is device 0 special?

twilk10 replied:
In my testing I have been getting device 0 often and came to the conclusion that it will be the most often used therefore the code bases getting the GPU powers off the assumption GPU 0 is where the information the user needs is.

masterleinad replied:
The reason why you see GPU 0 exclusively is that you will only use a different GPU if running with MPI (and multiple GPUs are present) or if you are requesting a different GPU explicitly at initialization, see https://kokkos.org/kokkos-core-wiki/ProgrammingGuide/Initialization.html?highlight=device+id#initialization-by-command-line-arguments.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -142,107 +195,201 @@ std::string variorum_print_power_call() {
+      }
+    }
+
+    if (gdevID == 0) {
+      switch (gdevID) {
+        case 1: {
```
This doesn't make sense. If you are only entering the if-body for device id 0, you will never execute any of the switch statements.

twilk10 replied:
I will fix this.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -142,107 +195,201 @@ std::string variorum_print_power_call() {
+    if (gdevID == 0) {
+      switch (gdevID) {
+        case 1: {
+          json_t* gpu_value = json_object_get(power_gpu_watts, "GPU_1");
```
```suggestion
          json_t* gpu_value = json_object_get(power_gpu_watts, "GPU_" + std::to_string(gdevID));
```
and remove all the switch cases.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -142,107 +195,201 @@ std::string variorum_print_power_call() {
+      temp = time2;
+
+      output = "  Energy Estimation " + std::to_string(((power1 + power2) / 2) *
+                                                       ((temp - time1) * .001));
```
What's the unit?

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -318,35 +465,11 @@ void kokkosp_init_library(const int loadSeq, const uint64_t interfaceVer,
     std::cout << "No MPI Option provided, not using per rank output"
               << std::endl;
-    usingMPI = false;
+    //  usingMPI = false;
```
```suggestion
```

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -356,83 +479,98 @@ void kokkosp_init_library(const int loadSeq, const uint64_t interfaceVer,
+
+  writeToFile(filename, output);
+  return os;
 }
```
Please don't introduce an overload for `operator<<` but use a `stringstream` or write to the file immediately instead.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -356,83 +479,98 @@ void kokkosp_init_library(const int loadSeq, const uint64_t interfaceVer,
+  writeToFile(filename, "name: " + std::string(name));
+  gdevID      = devID;
+  auto result = Kokkos::Tools::Experimental::identifier_from_devid(devID);
+  std::cout << result;
```
Please don't print to `cout` anywhere but only to the output file.

masterleinad commented:
> Sample output; `name: Z4mainEUllRlE_ Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation 2679.104000`

Please provide output from the incremental tests or so.

Commit:
remove unneeded files

Commit:
addressing comments on pull request

Commit:
addressing comments on pull request

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -26,13 +27,26 @@
+int power2      = 0;
+int filemake    = -1;
+long long time1 = 0;
+long long time2 = 0;
```
Can you be more precise?

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -26,13 +27,26 @@
+int filemake    = -1;
+long long time1 = 0;
+long long time2 = 0;
+// This is a global variable instituted to track what the device ID is
+uint32_t gdevID = -1;
```
You shouldn't need a global variable for that.

twilk10 replied:
Thank you for catching that.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -75,9 +89,36 @@ time_t start_time;
-bool verbosePrint = false;
-bool mpiOutPut    = false;
+
+bool verbosePrint = true;
```
Do you still need this? What does it do?

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -75,9 +89,36 @@ time_t start_time;
+}
+
+void printFile() {
+  // std::string filename = getFile("KOKKOS_TOOLS_VARIORUM_OUTPUT_FILE");
```
Why is this commented?

twilk10 replied:
To give the option to users to use environment variable if they chose to or have the file variorumoutput.txt put directly into their directory.

masterleinad replied:
Right but why is it not active? If you want to fall back to "variorumoutput.txt", you should not error out in `getFile` but rather use that if the environment variable couldn't be parsed correctly.

twilk10 replied:
That is a good idea

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -89,19 +130,19 @@ bool mpiOutPut    = false;
 std::string variorum_print_power_call() {
   std::string outputString;
-  json_t* power_obj = json_object();
+  char* s = NULL;
```
```suggestion
  char* s = nullptr;
```

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -142,107 +182,109 @@ std::string variorum_print_power_call() {
+    json_t* socket_0        = NULL;
+    json_t* timestamp_value = NULL;
+    json_t* power_gpu_watts = NULL;
+    json_t* gpu_0_value     = NULL;
```
```suggestion
    json_t* root            = nullptr;
    json_t* socket_0        = nullptr;
    json_t* timestamp_value = nullptr;
    json_t* power_gpu_watts = nullptr;
    json_t* gpu_0_value     = nullptr;
```

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -142,107 +182,109 @@ std::string variorum_print_power_call() {
+      fprintf(stderr, "Expected 'power_gpu_watts' to be an object.\n");
+    }
+
+    std::cout << "Device ID: " << gdevID << std::endl;
```
```suggestion
```

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -142,107 +182,109 @@ std::string variorum_print_power_call() {
+    } else {
+      power2 = gpu_0;
+    }
+    *
```
What happened here?
```suggestion
```

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -142,107 +182,109 @@ std::string variorum_print_power_call() {
+    }
   }
 }
-
```
Please revert deleting this line.

twilk10 replied:
Which line are you referring to.

masterleinad replied:
Look at the comments via https://github.com/kokkos/kokkos-tools/pull/265/files.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -318,121 +360,116 @@ void kokkosp_init_library(const int loadSeq, const uint64_t interfaceVer,
+          result.device_id));
+  writeToFile(filename, output);
+
+  variorum_call();
```
Pass the device id here.

twilk10 committed:
Update profiling/variorum-connector/variorum-connector.cpp

Co-authored-by: Daniel Arndt <arndtd@ornl.gov>

Commit:
update based on comments on pull request 10:48 8/8

Commit:
updates based off of pull requests 15:37 8/8

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -31,20 +31,28 @@
-int filemake    = -1;
-long long time1 = 0;
-long long time2 = 0;
+int power = -1;  // This variable is used to keep track of if variorum call has
+                 // been called twice to check if the variables need to be reset
+                 // for the next kernel
```
Wouldn't a `bool` be better?

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -31,20 +31,28 @@
+    0;  // This variable is used to obtain the initial power for calculation
+double power2 =
+    0;  // This variable is used to obtain the final power for calculation
+int filemake = -1;  // This variable is to check if the file has already been
+                    // made earlier in the program
```
Wouldn't a `bool` make more sense?

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -31,20 +31,28 @@
+int power = -1;  // This variable is used to keep track of if variorum call has
+                 // been called twice to check if the variables need to be reset
+                 // for the next kernel
+double power1 =
+    0;  // This variable is used to obtain the initial power for calculation
+double power2 =
+    0;  // This variable is used to obtain the final power for calculation
```
Could be a double array.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -31,20 +31,28 @@
+    0;  // This variable is used to obtain the final power for calculation
+int filemake = -1;  // This variable is to check if the file has already been
+                    // made earlier in the program
+long long time1 =
+    0;  // This variable is used to obtain the initial time for calculation
+long long time2 =
+    0;  // This variable is used to obtain the final time for calculation
```
Could be a double array.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -88,19 +96,20 @@ time_t start_time;
+    parsed_output_file = vararr;
   }
   return std::string(parsed_output_file);
 }
```
Wouldn't it be sufficient to call this when `Kokkos` is initialized?

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -88,19 +96,20 @@ time_t start_time;
-                 "variable!\n";
-    std::abort();
+                 "variable! Printed to variorumoutput.txt\n";
+    //  parsed_output_file = "variorumoutput.txt";
+    char vararr[19]    = "variorumoutput.txt";
+    parsed_output_file = vararr;
```
```suggestion
    return "variorumoutput.txt";
```

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -140,7 +149,7 @@ std::string variorum_print_power_call() {
   const char* hostnameChar =
       json_string_value(json_object_get(power_obj, "hostname"));
-
+}
```
This looks suspicious.

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -308,34 +304,15 @@ void kokkosp_init_library(const int loadSeq, const uint64_t interfaceVer,
-    
-  }
   // Simple timer code to keep track of the general amount of time the
   // application ran for.
```
Is any of this still used?

Commit:
updates based off of pull request comments 14:04 8/9

twilk10 commented:
Current out put with Incremental test
```name: RangePolicy_ParallelFor Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 231.882000
name: Reduction Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 270.228000
name: Reduction Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 232.134000
name: Reduction Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 233.646000
name: Reduction Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 235.074000
name: Reduction Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 233.436000
name: Reduction Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 232.008000
name: Reduction Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 234.276000
name: Reduction Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 231.882000
name: Reduction Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 232.050000
name: Reduction Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 232.176000
name: Reduction Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 231.462000
name: Reduction Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 233.394000
name: Reduction Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 234.108000
name: Reduction Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 234.528000
name: Reduction Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 231.840000
name: Reduction Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 231.462000
name: Reduction Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 231.378000
name: Reduction Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 233.520000
name: Reduction Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 233.646000
name: Reduction Device ID: 0 Instance ID: 1 DeviceType: SERIAL Energy Estimation in Joules 231.042000

```

Commit:
updates based off of pull request comments 14:20 8/9

masterleinad committed:
Clean up the variorum connector some more

masterleinad commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -14,425 +14,220 @@
-
-// variables for simple timer
-time_t start_time;
+// Initial and final power values for a kernel
+double global_power[2] = {0, 0};
+// Initial and final time value for a kernel
+long long global_time[2]    = {0, 0};
+uint32_t global_device_id   = -1;
+uint32_t global_instance_id = -1;
```
Obvisously, this tool is not thread-safe.

masterleinad committed:
Remove another unused variable

JBludau commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -14,425 +14,220 @@
+uint32_t global_instance_id = -1;
+std::string global_filename;
+std::string global_kernel_name;
+std::string global_device_type;
```
why do we need so many global variables?

masterleinad replied:
We need to record a bunch of information in the "begin kernel" calls that we are using in the "end kernel" calls. We can't pass that through the API interfaces directly.

janciesko commented:
Might be a good idea to move forward with this PR

masterleinad commented:
> Might be a good idea to move forward with this PR

A review would help. 🙂

dalg24 commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -14,425 +14,220 @@
-    variorum_call();
-  }
+  auto result = Kokkos::Tools::Experimental::identifier_from_devid(devID);
+  global_kernel_name = std::string(name);
```
```suggestion
  global_kernel_name = name;
```
Same comment everywhere

dalg24 commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -14,425 +14,220 @@
-  json_t* my_power_obj = NULL;
+void variorum_call() {
+  char* s              = nullptr;
+  json_t* my_power_obj = nullptr;
   my_power_obj         = json_object();
```
```suggestion
  json_t* my_power_obj = json_object();
```

janciesko replied:
Fragment from previous version, remove `my_power_obj`. We're using `root` now instead.

dalg24 commented on profiling/variorum-connector/variorum-connector.cpp:
```diff
@@ -14,425 +14,220 @@
+
+  // Parse JSON string into a json_t object
+  root = json_loads(s, 0, &error);
+  if (!root) {
+    fprintf(stderr, "Error parsing JSON: %s\n", error.text);
+  }
```
Consistency: use `std::cerr`

dalg24 reviewed (commented):
replying to https://github.com/kokkos/kokkos-tools/pull/265#issuecomment-2593687904

I will defer to you.

masterleinad committed:
Replace fprintf

masterleinad committed:
Don;t call std:string constructor unnecessarily

masterleinad committed:
Build variorum

janciesko commented:
Thanks @masterleinad for adding the CI testing. Seems to compile fine with Variorum for me.

janciesko commented:
I am ok with this if CI with Variorum succeeds.

masterleinad committed:
Simplify

masterleinad committed:
Fix formatting

masterleinad committed:
Install and set Fortran compiler

masterleinad committed:
Install Jansson

masterleinad committed:
Simplify installation

masterleinad committed:
Variorum_ROOT -> VARIORUM_ROOT

dalg24 commented on .github/workflows/build-with-kokkos.yml:
```diff
@@ -88,6 +104,7 @@ jobs:
           cmake --install build-with-${{ matrix.preset }} --prefix=${Kokkos_ROOT}
       - name: Build Kokkos Tools, enabling examples
         run: |
+          export CMAKE_PREFIX_PATH=${CMAKE_PREFIX_PATH};${VARIORUM_ROOT}
```
I expect the environment variable already adds it to the search path

dalg24 replied:
https://github.com/kokkos/kokkos-tools/blob/dd08b98dafb40f3291704eb54f20cf25cece3b78/cmake/configure_variorum.cmake#L19
Apparently it is spelled with an uppercase `V`.
Please adjust your env variable accordingly and remove the line I highlighted.

masterleinad replied:
The previous commit had `Variorum_ROOT` as environemnt variable and didn't find `variorum`. https://github.com/kokkos/kokkos-tools/blob/dd08b98dafb40f3291704eb54f20cf25cece3b78/cmake/configure_variorum.cmake#L8-L10 shows that it should be all uppercase.

dalg24 replied:
That configure cmake file logic is absurd and need fixing.
(On my cell can't navigate easily but in short it should not write to Variorum_ROOT that is nonsensical)

masterleinad committed:
Don't set CMAKE_PREFIX_PATH

masterleinad committed:
Fix typo

masterleinad committed:
Skip Variorum test

masterleinad committed:
Don't use GTEST_FILTER

dalg24 commented on .github/workflows/build-with-kokkos.yml:
```diff
@@ -92,7 +108,8 @@ jobs:
         if: ${{ matrix.preset == 'OpenMP' }}
         run: |
-          ctest --preset=${{ matrix.preset }}
+          ctest -E "test_kokkos_tools_variorum" --preset=${{ matrix.preset }}
```
Why are we excluding that test?

dalg24 commented on .github/workflows/build-with-kokkos.yml:
```diff
@@ -92,7 +108,8 @@ jobs:
           cmake --build --preset=${{ matrix.preset }}
           cmake --install build-with-${{ matrix.preset }} --prefix=/opt/kokkos-tools
       # For now, GitHub runners are used. These runner don't have GPUs. Therefore, testing can only be done for OpenMP.
+      # Skip variorum test since we can't rely on the hardware to compatible.
```
Fix sentence please

masterleinad replied:
```suggestion
      # Skip variorum test since we can't rely on the system to be compatible with the access Variorum needs.
```

masterleinad committed:
Improve comment

masterleinad commented on .github/workflows/build-with-kokkos.yml:
```diff
@@ -92,7 +108,8 @@ jobs:
           cmake --build --preset=${{ matrix.preset }}
           cmake --install build-with-${{ matrix.preset }} --prefix=/opt/kokkos-tools
       # For now, GitHub runners are used. These runner don't have GPUs. Therefore, testing can only be done for OpenMP.
+      # Skip variorum test since we can't rely on the system to be compatible with the access Variorum needs.
```
```suggestion
      # Skip variorum test since the architecture detected by variorum is not supported
```

masterleinad committed:
Improve comment again
